Show only latest VS, VC prompts by default

heaths opened the pull request:
## Summary of the Pull Request

Similar to `vswhere -latest`, show only the latest Visual Studio command prompts / developer PowerShell. This was tested by deleting the local package state and testing against fresh state with both VS2019 and VS2022 Preview installed, and indeed VS2022 Preview (both cmd and powershell) show. The other profiles were generated but hidden by default.

## References

Modification of PR #7774

## PR Checklist
* [x] Closes #11307
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

The sort algorithm is the same basic algorithm I used in https://github.com/microsoft/vswhere. It sorts first by installation version with a secondary sort based on the install date in case the installation versions are the same.

## Validation Steps Performed

With both VS2019 and VS2022 Preview installed, I made sure the initial state was expected, and tried different combinations of hiding and unhiding generated entries, and restarted Terminal to make sure my settings "stuck".

heaths commented:
@zadjii-msft though I see some of the other shells you saw for cross-compiling, etc., they are not being enumerated in my private build even before I made source modifications. Did you do anything to get them to show up?

![image](https://user-images.githubusercontent.com/1532486/134631822-21d17971-6657-47b6-8af2-cfcb5d3ff3cf.png)

I do have those other shells in my Start menu though:

![image](https://user-images.githubusercontent.com/1532486/134630622-5efa313d-78a7-4328-9039-9a4742a45ea6.png)

IIRC, the shortcut descriptions are localized but we could safely assume a pattern match of, effectively, `\bx64\b` for a native 64-bit WT process, and `\bx86\b` for x86. But if they don't even show up by default, the extra processing seems unnecessary and we could stick to just showing the latest by default as [discussed](https://github.com/microsoft/terminal/issues/11307#issuecomment-926036373).

heaths commented:
To note, those shortcuts are only installed if VC is installed, which means I have to either add a workload check, which will load a significantly large file and parse through it (the COM server will), or a file path check which will probably be faster. I'll need to spend a little time investigating when those variations were added (e.g. 15.0, or later in 15.3 or something) and whether the paths are consistent or need varying like @trippwill did for the PowerShell prompt since the path to the module changed.

If another preview is coming up soon, it might be worth taking this PR as-is to avoid spamming anyone with numerous VS installs, and I can open up another PR to continue investigating the aforementioned conditions.

zadjii-msft commented:
I appreciate taking the effort to break changes into smaller, more atomic pieces. I think we're targeting early Oct for the next release so we've still got time to iterate here, but you're right that this feels like something more easily digestible.

lhecker commented on src/cascadia/TerminalSettingsModel/VsSetupConfiguration.h:
```diff
@@ -37,6 +37,15 @@ namespace winrt::Microsoft::Terminal::Settings::Model
     public:
         struct VsSetupInstance
         {
+            VsSetupInstance(VsSetupInstance&& other) :
+                query(std::move(other.query)),
+                inst(std::move(other.inst)),
+                profileNameSuffix(std::move(other.profileNameSuffix)),
+                installDate(other.installDate),
+                version(other.version)
+            {
+            }
```
You can write `VsSetupInstance(VsSetupInstance&&) = default;` and move the assignment operator up to be next to it.

heaths replied:
That's what I had originally, but seems the default wasn't equivalent to what I have here (though it seems it should've been). It was causing dereferencing exceptions on shutdown.

heaths replied:
Before marking this resolved, I'll `git clean -dfx` my repo and start completely clean in case there was some issue with an incremental build (I built `main` before starting work to have a baseline). FWIW, incremental builds still took quite a while so I figured - per-project at least - they were recompiling all source and linking all new objects.

heaths replied:
The existing code wasn't properly AddRef'ing the `com_ptr<ISetupConfiguration2>`. Fixed, and the shutdown crash eliminated.

I'm also going to refactor this so it's only looping through instances once. It'll be more efficient for startup time, and make adding VC environments less costly.

lhecker commented on src/cascadia/TerminalSettingsModel/BaseVisualStudioGenerator.cpp:
```diff
@@ -11,10 +11,17 @@ void BaseVisualStudioGenerator::GenerateProfiles(std::vector<winrt::com_ptr<impl
+    static auto instances = VsSetupConfiguration::QueryInstances();
 
-    for (auto const& instance : instances)
+    // Sort instances based on version and install date.
+    std::sort(instances.begin(), instances.end(), BaseVisualStudioGenerator::Compare);
```
Could you move this into `VsSetupConfiguration::QueryInstances`?
That way this function would remain thread safe.

Additionally if you reverse the sort order you can iterate through it with your regular for loop.

Instead of defining an explicit function for `BaseVisualStudioGenerator::Compare` you can write a lambda:
```
std::sort(instances.begin(), instances.end(), [](const auto& a, const auto& b) {
    const auto aVersion = a.GetComparableVersion();
    const auto bVersion = b.GetComparableVersion();

    if (aVersion == bVersion)
    {
        return a.GetComparableInstallDate() > b.GetComparableInstallDate();
    }

    return aVersion > bVersion;
});
```

heaths replied:
I prototyped with a lambda, but this function is going to likely grow. Personally not a fan of monolithic lambdas but if you prefer that I will.

lhecker replied:
Please keep it as a separate function if you prefer that. 🙂 But I'd like to ask you to move it into `VsSetupConfiguration::QueryInstances`, as that would keep the function thread-safe. I believe reversing the sort order a good idea as well.

lhecker replied:
By the way: Do we need to use stable sort here? If so we should switch to `std::stable_sort`.

heaths replied:
It's practically impossible when comparing version + install date that you have equivalent instances. The VS setup engine can only install 1 instance at a time and blocks for others until it's finished. Not only did we do that to mitigate issues with machine state changes based on what we detected before execution of our install plan began (impossible to prevent), but some packages we install - like Windows Installer packages (though very few) - will block concurrent installs anyway. We do have retry loops for that, but it makes installs take longer.

As such, and given the relatively low number of instances expected on the vast majority of devices (from telemetry, 1 dwarfs 2 or more instances, and it quickly falls off after 2 and especially 3 instances) I don't think `std::stable_sort` is necessary.

I will move the sort inside `QueryInstances`, though. It's a good idea.

lhecker commented on src/cascadia/TerminalSettingsModel/VsSetupConfiguration.h:
```diff
@@ -167,6 +194,22 @@ namespace winrt::Microsoft::Terminal::Settings::Model
+                return VsSetupConfiguration::GetInstallDate(inst.get());
+            }
+
+            unsigned long long GetInstallationVersion() const
```
I'm confused what the difference between this `GetInstallationVersion` and `VsSetupConfiguration::GetInstallationVersion` is... Since they have the same name, shouldn't they also do the same thing?

heaths replied:
One is calling the equivalent method on the com_ptr to get and cache the value. The other is as const that simply returns the cached value. std::sort() is going to get the value multiple times so it's be inefficient to not only fetch the value but QI the setup helper.

heaths commented:
As discussed, only the latest and processor architecture-matching native tools environment (if VC++ is installed) are shown by default:

![image](https://user-images.githubusercontent.com/1532486/134799829-06bc88b0-7038-46ec-a0b2-74355a6eb7d7.png)

zadjii-msft commented:
Team discussion: Yea we're pretty cool with this, and the fact that you can mark generated profiles as hidden by default is really cool. We'll try and get this in for 1.12 ☺️

heaths commented:
If 1.11 is coming up, note that while the profile GUIDs shouldn't have changed from what @trippwill wrote, the namespace did since there's a single generator. Not sure if that's going to pose a migration problem.

That said, if there's time, I think we should at least take the sorting / hiding of all but the latest VS and can pull in the VC later. I could fairly easily remove the VC detection if you wanted to preview that longer and get feedback. With VS2022 coming out soon, more people will have multiple VSes installed (at least for a little while, based on past telemetry) so there will be at least 4 new entries all for VS. I worried about a backlash because of "spam".

zadjii-msft commented:
Oh, I don't think we'll have any troubles merging this for the next release, we still have a little under 2 weeks so that's plenty of time

heaths commented:
Isn't the original code already in 1.11 preview and getting close to shipping?

zadjii-msft commented:
@heaths I don't believe so, #7774 only merged on the 15th and [1.11 preview](https://github.com/microsoft/terminal/releases/tag/v1.11.2421.0) came out on August 31st

zadjii-msft reviewed (approved):
Yea I'm cool with this. I like that we're combining this into one namespace, that makes a lot of sense.

heaths commented on src/cascadia/TerminalSettingsModel/VcDevCmdGenerator.h:
```diff
@@ -0,0 +1,46 @@
+    private:
+        bool IsInstanceValid(const VsSetupConfiguration::VsSetupInstance&) const
+        {
+            // TODO: Need to determine when these VC environment scripts were added. vcvars32.bat at least was there since 15.0.
```
"TODO eventually". I'm working to find out, but how the code is written if those files aren't there, there'd be no error. I'll find out ASAP. If you want to merge this PR otherwise, I can follow it up with another as soon as I close on a decision with the VC team.

heaths replied:
I was able to track down the Dev 15.0 (latest service version) and indeed the path was the same then as it is in VS2019 (latest) and VS2022 Preview (latest). This path should be fine, then. I'll remove the TODO.

LuanVSO commented:
could/should it also prefer powershell 7 for devshell?

lhecker commented:
@LuanVSO Totally! If you don't mind, could you open a feature request issue for this?

DHowett commented:
Sorry, I'm just coming back from vacation and doing post-mortem PRs.

I read over the body of the PR and the committed changes, and I can't tell why _so much_ code changed. What else happened here? :smile:

heaths commented:
I refactored into a single discoverer because the existing ones were coded to generate a single profile per VS version. The VC cmd shell were to generate 1 or 4 depending on chip. Once I refactored, the inheritance model of the old generates was unnecessary and I was able to remove some otherwise unnecessary code. No point keeping it around to maintain.

DHowett commented:
Love it! Thank you
